nasa-gibs/worldview
# Layer grouping

## Description

Fixes #3185 

### What Changed

* add grouping for layers
* new permalink param added to indicate grouping enabled/disabled for each compare mode: `lg` and `lg1`
* `layergroup` property for all layer configs was converted to use the corresponding measurement for a layer and to be a single string value instead of an array
* a `tasks/generateLayergroup.js` config migration script is included to help migrate configs for other repos
* layer options/remove interactions now only show on hover
* some layer related selectors were simplified and/or memoized
* #3194 - revise event system to have all events go through a single instance. Update event names to be more descriptive.

NOTE: Some of these changes break parts of data download.  Since we are intending to deprecate data download in the same release that these changes are expected to be included, this shouldn't be an issue.  But, one data download e2e test appears to be failing and, if for some reason these changes are released without smart handoffs, any issues will need to be addressed.

### How to test

1. Open Worldview
2. Add some layers
3. Confirm layers get added to appropriate group
4. Re-order groups, confirm order is reflected on the map
5. Ungroup layers
6. Re-order layers arbitrarily
7. Confirm layer order is updated and reflected on the map
8. Turn grouping back on
9. Confirm layers are re-ordered into groups in the sidebar
10. Confirm layer order is updated and reflected on the map
11. Turn grouping off
12. Confirm arbitrary ordering is restored
13. Turn grouping back on
14. Remove any single layer
15. Turn grouping off
16. Confirm layer ordering now reflects ordering exhibited by grouped layers
17. Turn on compare mode
18. Turn grouping on for "A" side
19. Switch to "B" side
20. Confirm grouping is off
21. Try various combinations of adding/removing layers and groups on both "A" and "B" side and confirm both are properly maintained when switching between "A" and "B" 

* At various steps, try copying the URL and loading from the permalink in a new tab/window to confirm loading from a permalink works
